src/traits.rs

```diff
@@ -121,6 +121,22 @@ pub trait RootContext: Context {
     fn on_queue_ready(&mut self, _queue_id: u32) {}
 
     fn on_log(&mut self) {}
+
+    fn create_http_context(&self, _context_id: u32, _root_context_id: u32) -> Box<dyn HttpContext> {
+        Box::new(EmptyHttpContext)
+    }
+
+    fn create_stream_context(
+        &self,
+        _context_id: u32,
+        _root_context_id: u32,
+    ) -> Box<dyn StreamContext> {
+        Box::new(EmptyStreamContext)
+    }
+
+    fn get_type(&self) -> ContextType {
+        ContextType::RootContext
+    }
```

Comment:
if I had to choose I'd prefer the former (ie have a single `new_context` fn that creates whatever type), but at this point I think I like the more explicit approach more- because in the end we would have to be type-casting in the SDK again to call trait-specific fn's. but I'm open to suggestions